### PR TITLE
[Tabs] Adds justified and scrollable layout to MDCTabBarView

### PR DESCRIPTION
**Pull request by leonmz**
Adds justified and scrollable layout for MDCTabBarView and corresponding tests.

## Example View Controller

|Before|After (Portrait)|After (Landscape)
|---|---|---|
|60151483-ea246900-97aa-11e9-9c1e-688f1d241512|Simulator Screen Shot - iPhone 7 - 2019-06-26 at 08 27 47|Simulator Screen Shot - iPhone 7 - 2019-06-26 at 08 29 09|

Part of Issue #7645

### CODE REVIEW

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (mikefan1991) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: This is duplicate import.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (mikefan1991) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: This change is not necessary.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (mikefan1991) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: I have removed the title and the image in another PR. Please wait for that to be merged and then update this.

File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m
Comment (mikefan1991) [inline components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m]: This should stay commended.

**Review by mikefan1991 — CHANGES_REQUESTED** [review]

**Comment by romoore** [comment]
Gonna take a stab at cleaning this up. Ping me if you plan to work on it as well.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: This looks far too complex to be called from `-init`.  Let's try to simplify and add the constraints later.

**Review by romoore — CHANGES_REQUESTED** [review]

File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m
Comment (romoore) [reply components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m]: Done.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Reverted.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Removed.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Nit: Please document retention.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Please rename this to `commonMDCTabBarViewInit`.

**Review by romoore — CHANGES_REQUESTED** [review]

File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m
Comment (romoore) [inline components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m]: Please add a similar test that uses the bar's `intrinsicContentSize` to set the bounds so we can be sure it will return the correct size to fit the items.

File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m
Comment (romoore) [inline components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m]: Nit: Naming: Please remove "InScollView" from the test since the behavior is automatic and depends on bounds sizes, etc.  Consider the name `testItemsLargerThanBoundsChangesToScrollableLayout`?

File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m
Comment (romoore) [inline components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m]: A few things here: 
*  Please make the bounds larger than the single item's maximum width (I believe it's 360)
*  Please remove "InScrollView" from the name, since the layout depends on content + bounds and isn't guaranteed.  Perhaps `testVeryLongItemLimitsWidth` or something similar?

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Not sure if this change resolve this issue..

File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m
Comment (leonmz) [reply components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m]: Added.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Fixed.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Fix applied.

File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m
Comment (leonmz) [reply components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m]: Thanks for the advising!

File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m
Comment (leonmz) [reply components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m]: Added :)

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: I don't see a difference. Perhaps you need to push? I think it's also better to name this for its semantic use (rather than its type).  I think something like this would be better:

```objc
@property(nonnull, nonatomic, strong) UIStackView *containerView;
```

Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Fix applied!

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (mikefan1991) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: #pragma mark - Private API should be added for the following functions.

Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Fixed.

**Review by mikefan1991 — APPROVED** [review]

**Comment by leonmz** [comment]
I refactored the colorFortState function a bit. Deleted the items view and use the arranged view from containerView instead. Please take a look at it. Thanks!

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Please add a nullability annotation to this property. Please also use generics to document that it contains `NSLayoutConstraint *` types.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Please don't call `addConstraints` since it will be deprecated. Instead, activate the constraints directly.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m:275]: It seems like `layoutSubviews` shouldn't be required if a constraint-based layout is used. Even then, I would expect that this code execute before `[super layoutSubviews]` since that is expected to use any constraints to layout the subviews.

I suspect a better place to configure `self.containerView.distribution` might be in `setItems` and in `setBounds`.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: This isn't necessary since `removeFromSuperview` automatically removes it from the stack view's `arrangedSubviews` property.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Nit: Please use `CGRectGetWidth(self.bounds)` instead of accessing the bounds directly. Especially for UIScrollViews, the bounds origin may not be (0, 0).

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Looks like this can be replaced with a call to `self.intrinsicContentSize`?

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: I'm curious about this logic. If the current bounds is `CGRectZero` and `sizeToFit` is called, then this view would only resize its height. Is that expected?  Should it not attempt to provide a fit size for its contents as a minimum, but allow a larger width?

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: Nit: Please move this up near the `-init` methods.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: This call doesn't seem necessary.  I believe the `items` array is empty during initialization.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m]: I believe this logic can be put back in `setItems:` since it's only called from there.

**Review by romoore — CHANGES_REQUESTED** [review]
I'm going to address most of my feedback and push up changes to the branch.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Ah, no, that returns the scrollable width, not the justified width.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m:275]: I'll leave this here for now. `setBounds:` is called more times during the initial presentation of a view and not much less often during rotation. It's likely not worth the optimization.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: I still feel it worth a function extract the logic..

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done. I was able to remove this property and replace it with a flag. #namingIsHard

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Done. Moved into `-init`

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Removed.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Could you help me understand your thinking? We cannot unit test this code and it's only ever called from within `setItems:`. Either testability or duplicate code would be good candidates for a method. In this case I'm trying to understand the value of an extra message send (plus complexity in figuring out what the method does).

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: I am thinking about since it is a scrollable view thus this should be fits in any width.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m:291]: Done?

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Removed.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: CGZero is a extremely case, but if a small bounds is provided, the scrollable view should also fit itself in

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Thanks  a lot for detailed explanation!!

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: That's true, but the fit size would (ideally) be the size of its content. I think the best thing to conform to the documented behavior of `sizeThatFits:` is to return the "fit size".  In this case, it's probably the the size of the justified layout (rather than the intrinsic content size).  Does that make sense?

Whether it's kept like this or changed, I think it's worth adding a unit test that calls `sizeToFit:` with a bounds of `CGRectZero` and make sure it gets back a non-zero width and non-zero height.  Something similar for a starting bounds bigger than necessary?

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m:297]: Please add unit tests to cover verify the behavior/values of `intrinsicContentSize`.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [inline components/Tabs/src/TabBarView/MDCTabBarView.m:309]: I also mentioned it in another thread, but please add unit tests for `sizeThatFits:`

**Review by romoore — CHANGES_REQUESTED** [review]
Looks good, but needs unit tests for `intrinsicContentSize` and `sizeThatFits:` before it's ready for merge.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m:297]: Added.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m:309]: Added.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (leonmz) [reply components/Tabs/src/TabBarView/MDCTabBarView.m]: Sure. Fixed applied. I also refactored the intrisicSize to justified width instead of scrollable width.

**Comment by leonmz** [comment]
Refactored the intrsicSize to justified size. Also added snapshot tests for case that the bounds is 1000 but only two items exist. The current behavior is that the item view has been stretched to 500 due to its low hugging priority.

File: components/Tabs/src/TabBarView/MDCTabBarView.m
Comment (romoore) [reply components/Tabs/src/TabBarView/MDCTabBarView.m:297]: No, I don't think the TabBarView should be dependent on its contents.  I'd prefer the tests cover the more general behavior of `intrinsicContentSize`. For example, ensuring that it returns the justified/scrollable width consistently (depending on what you determine is more useful). That the height grows if one of its items has a very large image or long text.  It's the relationship/behavior that's important and not the exact values.

Let me add one or two as a commit and we can talk about them.  I think that will be faster and clearer than my trying to explain in a comment. ;-)

**Review by romoore — APPROVED** [review]

**Comment by romoore** [comment]
@googlebot All authors are Googlers, but one commit has a misconfigured author email.

**Pull request merged**
